### DISCUSSION TITLE
假如可以获取到所有目标点位，是否可以自动生成寻路路径而免去录制的过程？

### DISCUSSION THREAD

**Discussion by tignioj** [tignioj/minimap#11] [Announcements]
## 现状
目前脚本的寻路方式是按照已录制的路径进行寻路，人工录制的路线有重要的避障信息，包括途径点、行走方式。

## 手动录制方式寻路

优点：录制成本低，效果精准
缺点：需要大量人工

## 算法生成方式寻路
优点：无需人工录制路线

缺点：
1. 仍然需要人工添加避障信息，实际上并没有解决劳动力问题
3. 开发成本更高：需要额外维护一套传送点坐标
4. 目标点来源不明：需要一些特殊手段获取足够精准的材料点

## 结论
1. 如果人工录制作为寻路的最终答案，那么自动生成的寻路路径无论怎么优化都不如人工录制的效果
2. 仍然需要人工校准，并没有解决劳动力问题

**Comment by ttezy** [comment]
# 算法生成方式寻路的关键工作

## 地图处理
1. 用grid分割成小块，小块中心则是每次路径点的（x， y）
2. 利用PS人力涂色块，可行走路径涂白色，地图上的黑色边缘涂黑色或者灰色
3. 有水的地方可以用算法自动涂蓝色色块，但是人工涂应该也很快

## 算法寻路

平地可直接行走的cost为0，应填为白色块
较低可爬上的山体的cost会增加，应填为灰色块
太高爬不上的山体的cost为无穷大，应填为黑色块避免路径穿过
使用A* path search找到起点和终点间cost最低的路径

### 图例
<img width="801" alt="image" src="https://github.com/user-attachments/assets/71e3eab9-0d04-4293-8566-8cc4eba15fd1">

![image](https://github.com/user-attachments/assets/3942fea6-ea05-4eef-a68e-697a18567ad4)

### 地图边界的具体分类和cost还需要进一步讨论和细化

1. 从多高的山体开始从灰色块转为黑色块
2. 从灰色块的一侧到另一侧会跨越地形差，但有方向性，可能左侧到右侧是爬山，但是右侧到左侧则是飞行，爬山cost为高，但飞行cost很低，这样就需要定义单向通行的色块，可以用16个不同颜色来定义八个方位上的通行方向。或者用一种颜色比如红色代表可单向通行的边界，然后给每一个红色色块赋值（North/South/West/East/NW/NE/SW/SE）。前一种方法方便在PS里面涂色块操作。后一种方法比较好分类，相当于给了个二级菜单
3. 游泳可使用蓝色色块，游泳的cost则会根据游泳距离而增加，具体可参考准准的公式

### 准准的[视频1](https://www.bilibili.com/video/BV11P411d7fS/?spm_id_from=333.1296.top_right_bar_window_default_collection.content.click&vd_source=89fedb5955cfceadc1a60df93c76d55d)，[视频2](https://www.bilibili.com/video/BV1sY4y1X77q/?share_source=copy_web&vd_source=3601770997361fde771b5921a14e0f8f)
<img width="582" alt="image" src="https://github.com/user-attachments/assets/aa75e0fd-a8c7-4156-8e41-6eddc9244e19">
<img width="674" alt="image" src="https://github.com/user-attachments/assets/78748b1f-2cd6-4de9-b63b-3cb0a85341f4">
<img width="814" alt="image" src="https://github.com/user-attachments/assets/8b9bc50d-0b8f-456b-bad8-1eeb1586868d">

## 后续纠错
人工处理地图难免会出现涂错或者代码操作不过的地方，可能需要记录代码给出的path，然后更新方块的颜色（比如从可爬山的灰色方块update成黑色块）。

# 担忧
如果以上function都能实现的话应该是可以解决大部分路径点跟踪的问题的（我猜），但是如果做出来如此强大的功能不知道会不会被ban掉。。。

> 感谢辉鸭蛋的讨论和建议，以及提供的b站视频

**Comment by ttezy** [comment]
# Some additional research:
### [How to create a fictional map?](https://gis.stackexchange.com/questions/336056/how-to-make-a-fictional-city-map)
Use GIS or QGIS to create and labeling maps:

> GIS stands for Geographic Information System, and QGIS is a free, open-source GIS software: 

> GIS
> A system that organizes and visualizes geographic data, such as information about the Earth's surface, climate, and people. GIS can help people and businesses understand patterns and relationships in the world. 

> QGIS
> A GIS software that allows users to: 
> - View, analyze, and edit spatial data 
> - Create maps 
> - Manage geographic information 
> - Support a variety of spatial data file extensions, including .shp, .tif, .csv, and .img 
> - Be used on Windows, macOS, and Linux 
> - Provide a robust set of processing and analytical tools 
> - Have over 300 plugins that can be installed to provide additional functionality 
> - Be developed by contributors from around the world who are constantly adding new features and improving the program 

[Analyze data in QGIS using python
](https://www.qgistutorials.com/en/docs/getting_started_with_pyqgis.html)

> QGIS was previously known as Quantum GIS.

### 也许可以用现成工具做出来像2077/大表哥/GTA一样的大世界导航🤔？然后附加像2077 mod一样的游戏内导航辅助？
![image](https://github.com/user-attachments/assets/6c6dd675-9022-4e54-83d0-ba517c9c7f16)

https://github.com/user-attachments/assets/b9cd6c47-07ea-4e70-8b41-c950ecb02c2f

**Comment by ttezy** [comment]
# Update:
尝试了一下QGIS，在里面画出来路径点之后可以任选两点计算最短路径，这里我只画了能走的路，没考虑爬山游泳和飞行，相当于把地图拍扁了画的街道。用了QGIS自带的路径计算功能算了从蒙德城出口到摘星崖的路。

![image](https://github.com/user-attachments/assets/8166e5f4-1e6e-425b-9945-38f63882e626)

把路径点export成geopackage的格式（gpkg）可以用python读取路径点然后计算。我在python里尝试了同样的起始点和终点，得到的图长这样：

![path_search_example](https://github.com/user-attachments/assets/6f1be8e2-181c-4ffa-9af8-c78480418620)

# 问题：
### 设置坐标系原点
现在需要设置坐标系的中心点和地图的比例尺，目前QGIS的蒙德城（0，0）点是在地图图片左上角。我看 `\sifttest` 里写的以璃月为中心点但另一个地方写的以每个国家鼠标点击给出来的默认中心为中心点。所以不太明白怎么**设置坐标系**。

![image](https://github.com/user-attachments/assets/e3025cd4-1431-4036-9735-f09d01179cd3)

### 设置比例尺
另外就是比例尺的问题，我上面demo的两个点位是

```
start_point = Point(3127, -1685)  # Replace with your actual start UTM coordinates
end_point = Point(4670, -912)    # Replace with your actual end UTM coordinates
```
**怎么设置比例尺才能和minimap的参照系一样？**

**Reply by tignioj** [reply]
假如使用蒙德地图，先用sift直接计算出指定位置的坐标，该坐标默认为距离图片左上角的坐标，一个像素为一个单位。然后加上偏差就得到距离璃月地图设置的中心点（手动设置的天衡山右边的圆环，不是指点击地图后的中心点，后者为合成台锚点的坐标）的坐标。

![image](https://github.com/user-attachments/assets/864cfc9a-347f-4767-b733-502ba1404ffa)



每张图的和璃月地图设置中心点的偏差可以通过ps查看，具体的值已经在https://github.com/tignioj/minimap/blob/main/config.map.yaml 给出。

例如，选择了分割后的蒙德地图，在这里sift计算得到坐标是  ` (4503.03857421875, 754.2261962890625)`
![image](https://github.com/user-attachments/assets/36a61b8b-0298-4aa3-8991-cf490adaa1d0)

加上偏差
```
'蒙德':
  img_name: 'mengde'
  sigma: 1.4  # sift参数
  center: [427, 7359]  # 该图左上角坐标与璃月中心点坐标的偏差
  version: 0
```
即
```
x = 4503 - 427
y = 754 - 7359
得到minima最终坐标大概为
(4076.02001953125, -6604.75140380859
```
